islandora.tokens.inc

```diff
@@ -0,0 +1,124 @@
+<?php
+
+/**
+ * @file
+ * Contains islandora.tokens.inc.
+ *
+ * This file provides islandora tokens.
+ */
+
+use Drupal\media\Entity\Media;
+use Drupal\file\Entity\File;
+use Drupal\Core\Render\BubbleableMetadata;
+use Drupal\islandora\IslandoraUtils;
+
+/**
+ * Implements hook_token_info().
+ */
+function islandora_token_info() {
+  $type = [
+    'name' => t('Islandora Tokens'),
+    'description' => t('Tokens for Islandora objects.'),
+  ];
+  $node['media-thumbnail-image:url'] = [
+    'name' => t('Media: Thumbnail Image URL.'),
+    'description' => t('URL of Thumbnail Image associated with Islandora Object via Media.'),
+  ];
+
+  $node['media-thumbnail-image:alt'] = [
+    'name' => t('Alternative text for Media: Thumbnail Image.'),
+    'description' => t('Alternative text for Thumbnail Image associated with Islandora Object via Media.'),
+  ];
+
+  // Deprecated in favour if hyphenated version.
+  $node['media_thumbnail_image:url'] = [
+    'name' => t('Media: Thumbnail Image URL.'),
+    'description' => t('Deprecated: URL of Thumbnail Image associated with Islandora Object via Media.'),
+  ];
+
+  // Deprecated in favour if hyphenated version.
+  $node['media_thumbnail_image:alt'] = [
+    'name' => t('Alternative text for Media: Thumbnail Image.'),
+    'description' => t('Deprecated: Alternative text for Thumbnail Image associated with Islandora Object via Media.'),
+  ];
+
+
+  return [
+    'types' => ['islandoratokens' => $type],
+    'tokens' => ['islandoratokens' => $node],
+  ];
+}
+
+/**
+ * Implements hook_tokens().
+ */
+function islandora_tokens($type, $tokens, array $data, array $options, BubbleableMetadata $bubbleable_metadata) {
+  $replacements = [];
+  if ($type == 'islandoratokens' && !empty($data['node'])) {
+    if (!is_array($tokens) || empty($tokens)) {
+      \Drupal::logger('controlled_access_terms')
+        ->alert('Tokens not correct format: @tokens', [
+          '@tokens' => print_r($tokens, 1),
+        ]);
+      return;
+    }
+    foreach ($tokens as $name => $original) {
+      switch ($name) {
+        case 'media-thumbnail-image:url':
+        case 'media_thumbnail_image:url':
+          $term = $islandoraUtils->getTermForUri('http://pcdm.org/use#ThumbnailImage');
+          $media = $islandoraUtils->getMediaWithTerm($data['node'], $term);
+          // Is there media?
+          // @todo: is this single or multiple?
+          if ($media) {
+            $file = controlled_access_terms_image_from_media($media);
```

Comment:
you can use `$islandoraUtils->getSourceFile($media)` here and ditch the helper function